### PR TITLE
Add KeywordString to SynConst.

### LINKED ISSUE #11477: Add KeywordString to SynConst type
**Is your feature request related to a problem? Please describe.**

Any original keyword string (`__LINE__`, `__SOURCE_DIRECTORY__` or `__SOURCE_FILE__`) information is currently lost in the untyped tree.

They are represented as SynConst.String with the resolved values.
```fsharp
__LINE__
```
leads to 
```fsharp
Const(String ("1", tmp.fsx (1,0--1,8) IsSynthetic=false), tmp.fsx (1,0--1,8) IsSynthetic=false),
```

**Describe the solution you'd like**
I propose to extend `SynConst` with a new union case `| KeywordString of constant: string * value: string * range: Range`.

**Describe alternatives you've considered**
Get the original keyword from the F# tokens.

**Additional context**

This would lead to a richer AST and would be one less check Fantomas has to do to restore the original typed value.

### PR DESCRIPTION
**Pull request by nojaf**
Fixes #11477.

### CODE REVIEW

**Review by cartermp — APPROVED** [review]
Thanks!

**Comment by KevinRansom** [comment]
@nojaf -- KeywordString, doesn't really mean anything to me, and naming is the hardest problem in computer science.

Would perhaps SourceIdentifier be better, I note that the documentation for C# calls those values: source-line-file-path-identifiers
https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/source-line-file-path-identifiers

Just a thought, I really don't want to bikeshed.

**Comment by nojaf** [comment]
Hey Kevin, I'm totally ok with a different name.

File: tests/FSharp.Compiler.Service.Tests/SurfaceArea.netstandard.fs [hunk 1]
```diff
@@ -5778,6 +5778,12 @@ FSharp.Compiler.Syntax.SynConst+Int64: Int64 Item
 FSharp.Compiler.Syntax.SynConst+Int64: Int64 get_Item()
 FSharp.Compiler.Syntax.SynConst+IntPtr: Int64 Item
 FSharp.Compiler.Syntax.SynConst+IntPtr: Int64 get_Item()
+FSharp.Compiler.Syntax.SynConst+KeywordString: FSharp.Compiler.Text.Range get_range()
```
Comment (KevinRansom) [inline tests/FSharp.Compiler.Service.Tests/SurfaceArea.netstandard.fs:5781]: API baseline needs updating with the new name

File: tests/FSharp.Compiler.Service.Tests/SurfaceArea.netstandard.fs [hunk 2]
```diff
@@ -5804,6 +5810,7 @@ FSharp.Compiler.Syntax.SynConst+Tags: Int32 Int16
 FSharp.Compiler.Syntax.SynConst+Tags: Int32 Int32
 FSharp.Compiler.Syntax.SynConst+Tags: Int32 Int64
 FSharp.Compiler.Syntax.SynConst+Tags: Int32 IntPtr
+FSharp.Compiler.Syntax.SynConst+Tags: Int32 KeywordString
```
Comment (KevinRansom) [inline tests/FSharp.Compiler.Service.Tests/SurfaceArea.netstandard.fs:5813]: API baseline needs updating with the new name

File: tests/FSharp.Compiler.Service.Tests/SurfaceArea.netstandard.fs [hunk 3]
```diff
@@ -5860,6 +5868,7 @@ FSharp.Compiler.Syntax.SynConst: Boolean get_IsInt16()
 FSharp.Compiler.Syntax.SynConst: Boolean get_IsInt32()
 FSharp.Compiler.Syntax.SynConst: Boolean get_IsInt64()
 FSharp.Compiler.Syntax.SynConst: Boolean get_IsIntPtr()
+FSharp.Compiler.Syntax.SynConst: Boolean get_IsKeywordString()
```
Comment (KevinRansom) [inline tests/FSharp.Compiler.Service.Tests/SurfaceArea.netstandard.fs:5871]: API baseline needs updating with the new name

File: tests/FSharp.Compiler.Service.Tests/SurfaceArea.netstandard.fs [hunk 4]
```diff
@@ -5881,6 +5890,7 @@ FSharp.Compiler.Syntax.SynConst: FSharp.Compiler.Syntax.SynConst NewInt16(Int16)
 FSharp.Compiler.Syntax.SynConst: FSharp.Compiler.Syntax.SynConst NewInt32(Int32)
 FSharp.Compiler.Syntax.SynConst: FSharp.Compiler.Syntax.SynConst NewInt64(Int64)
 FSharp.Compiler.Syntax.SynConst: FSharp.Compiler.Syntax.SynConst NewIntPtr(Int64)
+FSharp.Compiler.Syntax.SynConst: FSharp.Compiler.Syntax.SynConst NewKeywordString(System.String, System.String, FSharp.Compiler.Text.Range)
```
Comment (KevinRansom) [inline tests/FSharp.Compiler.Service.Tests/SurfaceArea.netstandard.fs:5893]: API baseline needs updating with the new name

File: tests/FSharp.Compiler.Service.Tests/SurfaceArea.netstandard.fs [hunk 5]
```diff
@@ -5903,6 +5913,7 @@ FSharp.Compiler.Syntax.SynConst: FSharp.Compiler.Syntax.SynConst+Int16
 FSharp.Compiler.Syntax.SynConst: FSharp.Compiler.Syntax.SynConst+Int32
 FSharp.Compiler.Syntax.SynConst: FSharp.Compiler.Syntax.SynConst+Int64
 FSharp.Compiler.Syntax.SynConst: FSharp.Compiler.Syntax.SynConst+IntPtr
+FSharp.Compiler.Syntax.SynConst: FSharp.Compiler.Syntax.SynConst+KeywordString
```
Comment (KevinRansom) [inline tests/FSharp.Compiler.Service.Tests/SurfaceArea.netstandard.fs:5916]: API baseline needs updating with the new name

**Review by KevinRansom — APPROVED** [review]
Once the baseline is updated this will be good to go.

Thanks

**Comment by nojaf** [comment]
Thanks, forgot about SurfaceAreaTest indeed.

**Comment by KevinRansom** [comment]
@nojaf , thanks mate, it gets me every single time :-)

**Comment by alfonsogarciacaro** [comment]
After updating Fable with latest FCS code I'm getting problems when passing `__SOURCE_DIRECTORY__` as static parameter. Example:

```fsharp
open Zanaptak.TypedCssClasses

type Css = CssClasses<"./style.module.css", Naming.PascalCase, resolutionFolder=__SOURCE_DIRECTORY__>
```

Error in Fable logs:

```
Invalid static argument to provided type. Expected an argument of kind 'string'. (code 3045)
```

Can it be related to this?

**Comment by nojaf** [comment]
Possibly, the AST for this does indeed use the new SynConst.
```fsharp
StaticConstantNamed
                              (LongIdent
                                 (LongIdentWithDots ([resolutionFolder], [])),
                               StaticConstant
                                 (SourceIdentifier
                                    ("__SOURCE_DIRECTORY__",
                                     "/workspace/fantomas-tools/src/server/ASTViewer/bin/output",
                                     tmp.fsx (3,80--3,100) IsSynthetic=false),
                                  tmp.fsx (3,80--3,100) IsSynthetic=false),
                               tmp.fsx (3,63--3,100) IsSynthetic=false)
```

**Comment by alfonsogarciacaro** [comment]
Ah, sorry. I just realized this was fixed by #11610, we just need to pull the latest changes in Fable.

**Pull request merged**